curvefs/src/metaserver/metastore.cpp

```diff
@@ -0,0 +1,521 @@
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * @Project: curve
+ * @Date: 2021-08-30 19:43:26
+ * @Author: chenwei
+ */
+#include "curvefs/src/metaserver/metastore.h"
+#include <glog/logging.h>
+#include <thread>  // NOLINT
+#include <unordered_map>
+#include <vector>
+#include "curvefs/src/metaserver/storage.h"
+namespace curvefs {
+namespace metaserver {
+MetaStore::MetaStore() {}
+
+bool MetaStore::Load(const std::string& path) {
+    auto callbackFunc = [&](ENTRY_TYPE type, uint32_t paritionId,
+                            void* metadata) -> bool {
+        if (type == ENTRY_TYPE::PARTITION) {
+            PartitionInfo* partition =
+                reinterpret_cast<PartitionInfo*>(metadata);
+            partitionMap_.emplace(partition->partitionid(),
+                                  std::make_shared<Partition>(*partition));
+            return true;
+        } else if (type == ENTRY_TYPE::INODE) {
+            std::shared_ptr<Partition> partition = GetPartiton(paritionId);
+            if (partition == nullptr) {
+                LOG(ERROR) << "MetaStore Load GetPartiton fail, type INODE";
+                return false;
+            }
+            Inode* inode = reinterpret_cast<Inode*>(metadata);
+            MetaStatusCode status = partition->InsertInode(*inode);
+            return (status == MetaStatusCode::OK) ? true : false;
+        } else if (type == ENTRY_TYPE::DENTRY) {
+            std::shared_ptr<Partition> partition = GetPartiton(paritionId);
+            if (partition == nullptr) {
+                LOG(ERROR) << "MetaStore Load GetPartiton fail, type DENTRY";
+                return false;
+            }
+            Dentry* dentry = reinterpret_cast<Dentry*>(metadata);
+            MetaStatusCode status = partition->CreateDentry(*dentry);
+            if (status != MetaStatusCode::OK) {
+                LOG(ERROR) << "MetaStore Load CreateDentry fail, type DENTRY";
+                return false;
+            }
+            return true;
+        } else {
+            LOG(ERROR) << "load meta from file, not supported type";
+            return false;
+        }
+        return true;
+    };
+
+    // Load from raft snap file to memory
+    WriteLockGuard writeLockGuard(rwLock_);
+    bool ret = LoadFromFile(path, callbackFunc);
+    if (!ret) {
+        partitionMap_.clear();
+    }
+    return ret;
+}
+
+void MetaStore::SaveBack(const std::string& path, OnSnapshotSaveDone* done) {
+    // 1. call back dump function
+    std::vector<std::shared_ptr<Iterator>> iteractorList;
+    LOG(INFO) << "Save meta to file back.";
+    // add partition iteractor
+    std::map<uint32_t, PartitionInfo> partitionInfoMap;
+    for (auto partitionIter : partitionMap_) {
```

Comment:
add referenct to partitionIter to avoid copy,  `for (auto& partitionIter...`

Reply:
done